Add keypair generation on HSMs

istepic opened the pull request:
Introduced keypair generation on HSMs through the OpenSSL's ENGINE interface.
This is based on a temporary solution where libp11 is patched to support keypair
generation.

istepic committed:
Add hsm keygen

istepic committed:
update changelog

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
     std::string pin("1234");
     std::string keyID("1001");
```
This is no longer needed in this example.

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
     std::string pin("1234");
     std::string keyID("1001");
```
This is no longer needed in this example.

wusto replied:
Done

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
     HsmEngine hsmEngine(id, modulePath, pin);
 
-    // Step 4: Load the key from the HSM and print it.
-    auto pubKey = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyID);
```
Why not create a new example, in addition to this existing one, so that we still keep an example of using readPublicKeyFromHSM() for key loading?

Maybe create an file: "examples/hsm-gen-example.cpp", and include your example there?

garw replied:
Yep. I also think adding this on top would be better. I'm fine with reusing this file though - it can demonstrate several things.

istepic replied:
Done in the same file to demonstrate the API in one place.

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
+    ecdsaVerify(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, signature, message);
+    /*********************************************/
 
     return 0;
```
Nit: Maybe output something to stdout indicating success if all goes well?

wusto replied:
We don't have this for our other examples. Usually everything crashes when an exception is raised.

johnfxgalea-e commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
     return _ENGINE_load_private_key(_engine.get(), keyID);
 }
 
+openssl::SSL_EVP_PKEY_Ptr HsmEngine::genKeyGetPublic(const RSASpec &spec,
```
Keep function names simple and intuitive. s/genKeyGetPublic/genPublicKey

istepic replied:
Done

johnfxgalea-e commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+    return loadPublicKey(keyID);
+}
+
+openssl::SSL_EVP_PKEY_Ptr HsmEngine::genKeyGetPublic(const ECCSpec &spec,
```
Ditto

istepic replied:
Done

johnfxgalea-e commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+    return loadPublicKey(keyID);
+}
+
+openssl::SSL_EVP_PKEY_Ptr HsmEngine::genKeyGetPrivate(const RSASpec &spec,
```
Ditto

istepic replied:
Done

johnfxgalea-e commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+    return loadPrivateKey(keyID);
+}
+
+openssl::SSL_EVP_PKEY_Ptr HsmEngine::genKeyGetPrivate(const ECCSpec &spec,
```
Ditto

istepic replied:
Done

johnfxgalea-e commented on src/key.cpp:
```diff
@@ -86,7 +86,7 @@ AsymmetricKeypair AsymmetricKeypair::generateEd25519()
 }
 
-AsymmetricKeypair AsymmetricKeypair::generate(const AsymmetricKey::Spec &keySpec)
+AsymmetricKeypair AsymmetricKeypair::generate(const AsymmetricKey::Spec& keySpec)
```
Why are we getting these weird code changes all of a sudden? Is clang-format acting funny?

Seems like it might be due to DerivePointerAlignment, which precedes PointerAlignment and instead tries to determine the preferred alignment by analysing the file.

This needs to be fixed in a separate PR. I'll handle it.

johnfxgalea-e commented on src/key.cpp:
```diff
@@ -107,15 +107,35 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string
     return AsymmetricPublicKey{std::move(key)};
 }
+
+AsymmetricPublicKey AsymmetricPublicKey::genKeyOnHsmGetPublic(HSM& hsm,
```
s/genKeyOnHsmGetPublic/genPublicKeyOnHsm

istepic replied:
Done

johnfxgalea-e commented on src/key.cpp:
```diff
@@ -107,15 +107,35 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string
+    return AsymmetricPublicKey{std::move(key)};
+}
+
+AsymmetricPublicKey AsymmetricPublicKey::genKeyOnHsmGetPublic(HSM& hsm,
```
Ditto

istepic replied:
Done

johnfxgalea-e commented on src/key.cpp:
```diff
@@ -186,11 +206,31 @@ AsymmetricKeypair AsymmetricKeypair::readPrivateKeyFromPEM(const std::string &pe
     return AsymmetricKeypair{std::move(key)};
 }
+
+AsymmetricKeypair AsymmetricKeypair::genKeyOnHsmGetPrivate(HSM& hsm,
```
s/genKeyOnHsmGetPrivate/genPrivateKeyOnHsm

istepic replied:
Done

johnfxgalea-e commented on src/key.cpp:
```diff
@@ -186,11 +206,31 @@ AsymmetricKeypair AsymmetricKeypair::readPrivateKeyFromPEM(const std::string &pe
+    return AsymmetricKeypair{std::move(key)};
+}
+
+AsymmetricKeypair AsymmetricKeypair::genKeyOnHsmGetPrivate(HSM& hsm,
```
Ditto

istepic replied:
Done

johnfxgalea-e commented on src/mococrw/key.h:
```diff
@@ -92,6 +92,18 @@ class AsymmetricPublicKey
      */
     static AsymmetricPublicKey readPublicKeyFromHSM(HSM& hsm, const std::string& keyID);
+
+    static AsymmetricPublicKey genKeyOnHsmGetPublic(HSM& hsm,
```
Kindly add doxygen comments for this function, and the other new ones.

istepic replied:
Done

johnfxgalea-e commented on tests/unit/openssl_lib_mock.h:
```diff
@@ -35,6 +35,13 @@ namespace openssl
 class OpenSSLLibMockInterface
 {
 public:
+    virtual const char* SSL_EC_curve_nid2nist(int nid) = 0;
```
Where are the actual tests that use the mock? I would expect them to be added in MoCOCrW/tests/unit/test_hsm.cpp

istepic replied:
All code paths are covered by the examples. That's why I initially decided not to write unit tests, especially since example tests the functionality E2E and has more chance of reporting a problem.

istepic replied:
Added some unit tests to test_opensslwrapper.cpp

johnfxgalea-e reviewed (changes requested):
Main Points:
1) Add Doxygen comments
2) Include mock tests
3) Make function names more concise and clear
4) Add new example, rather than replacing existing one.

There is an issue concerning clang-format and pointer alignment. I will fix this in the meantime.

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
+    /************** ECDSA signature **************/
+    auto ecdsaDigestType = DigestTypes::SHA512;
+    ECCSpec ecspec;
+    auto eccPrivKey = AsymmetricPrivateKey::genKeyOnHsmGetPrivate(
```
In addition, also include an RSA example for completeness? (I know EC is more important, but would be nice to have both, especially since these examples also serve as additional tests).

wusto replied:
I guess that one example is enough here. Otherwise this is lots of copy and paste from other example.cpp's.

I added some lines how to use it with RSA.

garw commented on src/mococrw/key.h:
```diff
@@ -92,6 +92,18 @@ class AsymmetricPublicKey
      */
     static AsymmetricPublicKey readPublicKeyFromHSM(HSM& hsm, const std::string& keyID);
+
+    static AsymmetricPublicKey genKeyOnHsmGetPublic(HSM& hsm,
```
Keep the interface the same as for regular keys.
generateOnHsm(keyspec) or generateEccOnHsm(spec) generateRsaOnHsm(spec)
they always return a a private key. If you want to get a public key, you get this from the private key object. I don't see why the HSM should a have special generate method here that returns public keys. It is not there for normal keys and not needed.

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -92,6 +92,18 @@ class AsymmetricPublicKey
      */
     static AsymmetricPublicKey readPublicKeyFromHSM(HSM& hsm, const std::string& keyID);
+
+    static AsymmetricPublicKey genKeyOnHsmGetPublic(HSM& hsm,
```
Why do we have a key generation method in the pubkey class? From a logical point of view, only a keypair can be created, not a pubkey.

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -271,6 +283,17 @@ class AsymmetricKeypair : public AsymmetricPublicKey
      */
     static AsymmetricKeypair readPrivateKeyFromHSM(HSM& hsm, const std::string& keyID);
+
+    static AsymmetricKeypair genKeyOnHsmGetPrivate(HSM& hsm,
```
s.a.

istepic replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1597,5 +1597,16 @@ void _ENGINE_finish(ENGINE *e)
+
+std::string _EC_curve_nid2nist(int nid)
+{
+    return std::string{lib::OpenSSLLib::SSL_EC_curve_nid2nist(nid)};
```
We probably want error handling here and throw if the lookup wasn't successful?

To make things worse: std::string{nullptr} is undefined behavior, see https://en.cppreference.com/w/cpp/string/basic_string/basic_string case (5)

istepic replied:
Done

garw commented on src/mococrw/hsm.h:
```diff
@@ -53,6 +55,36 @@ class HSM
+                                                       const std::string &tokenLabel,
+                                                       const std::string &keyLabel) = 0;
+
+    virtual void genKey(const RSASpec &spec,
```
same as for the key objects, We don't need to abbreviate names. This should be called `generateKey`

istepic replied:
Done

garw commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+                       const std::string &tokenLabel,
+                       const std::string &keyLabel)
+{
+    PKCS11_RSA_KGEN rsa;
```
please more verbose names, this looks like `pkcs11_rsa_spec`

wusto replied:
Done

garw commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+{
+    PKCS11_RSA_KGEN rsa;
+    rsa.bits = spec.numberOfBits();
+    PKCS11_KGEN_ATTRS kg;
```
s.a.

wusto replied:
Done

garw commented on src/hsm.cpp:
```diff
@@ -45,4 +47,75 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::loadPrivateKey(const std::string &keyID)
+                       const std::string &tokenLabel,
+                       const std::string &keyLabel)
+{
+    PKCS11_EC_KGEN ec;
```
s.a.

wusto replied:
Done

garw commented on src/mococrw/hsm.h:
```diff
@@ -53,6 +55,36 @@ class HSM
      */
     virtual openssl::SSL_EVP_PKEY_Ptr loadPrivateKey(const std::string &keyID) = 0;
+
+    virtual openssl::SSL_EVP_PKEY_Ptr genKeyGetPublic(const RSASpec &spec,
```
This should have doxygen user documentation here. What do the parameters mean? What kind of input is allowed? How does this translate into PKCS#11 as we have PKCS#11 in mind. What would we do with it otherwise? How do we explain that the load functions just support a key id? Should it maybe support more? At least the label?

istepic replied:
Done. Still wondering if error reporting makes sense. Lots of things happen in libp11 and there's no clean way to pinpoint what went from MoCOCrW.

garw commented on src/mococrw/hsm.h:
```diff
@@ -53,6 +55,36 @@ class HSM
+                                                       const std::string &tokenLabel,
+                                                       const std::string &keyLabel) = 0;
+
+    virtual void genKey(const RSASpec &spec,
```
The normal key generation flow is, generate and get the private key. Why do we have these void functions and the other special get variants?

istepic replied:
Attempt to remove code duplication. Not needed anymore as API is simplified

garw commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,94 @@
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit upcast here.
+     */
+    ecdsaVerify(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, signature, message);
```
What I'm missing here is: get the private key and write out the public key as pem. Then reload it from PEM and use this for validation. This is what will (implicitly) happen a lot of times on our platforms.

wusto replied:
Done

garw commented on tests/unit/openssl_lib_mock.h:
```diff
@@ -35,6 +35,13 @@ namespace openssl
 class OpenSSLLibMockInterface
 {
 public:
+    virtual const char* SSL_EC_curve_nid2nist(int nid) = 0;
```
I agree here. We currently don't have any tests for the HSM code except for the example - which (iirc) also executed as a test but maybe a bit misleading.
This is actually a good place where this mock layer here comes in handy. You can actually write a test for your HSM class by mocking the openssl interface and then checking that is submits the right call to SSL_ENGINE_ctrl_cmd and even verify if the parameters look sane-ish.

istepic replied:
Done-ish

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1597,5 +1597,16 @@ void _ENGINE_finish(ENGINE *e)
+            lib::OpenSSLLib::SSL_ENGINE_ctrl_cmd, e, cmdName.c_str(), 0, p, nullptr, 1);
+}
+
+std::string _EC_curve_nid2nist(int nid)
```
This function may be interesting also to others. Maybe we want to offer this also in the public interface as static method on ECC Keyspec?

istepic replied:
Done

Commit:
Fix some of the review comments

istepic committed:
addressed some comments:
    - removed functions that generate and return public key
    - make all key generation function return private key by default
      to avoid many permutations of the same exact function
    - added doxygen comments
    - added error handling for _EC_curve_nid2nist

fixed constness in loadPrivateKey/loadPublicKey functions

istepic committed:
added some unit tests

istepic committed:
Add curveName() to ECCSpec.

This method returns NIST name of the ECC curve as string.

istepic committed:
added additional unit test

- introduced _EVP_PKEY_bits to make it compile

istepic committed:
refactoring
  - templated generateKeyOnHsm function
  - added explicit checks for validity of keyID

johnfxgalea-e commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,123 @@
+    auto signature = ecdsaSign(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, message);
+
+    /* Use-Case 1: You want to check your own signature:
+     * we can use here the private key, as it also contains the public key.
```
Nit: s/w/W

istepic replied:
Done

istepic committed:
remove unecessary code

hsm-example no longer uses keys pregenerated on softhsm
using pkcs11-tool. Instead, it uses keys that are generated on
softhsm using HsmEngine API.

johnfxgalea-e commented on src/mococrw/hsm.h:
```diff
@@ -64,6 +94,16 @@ class HsmEngine : public HSM
     HsmEngine(const std::string &id, const std::string &modulePath, const std::string &pin);
     virtual ~HsmEngine();
 
+    openssl::SSL_EVP_PKEY_Ptr generateKey(const RSASpec &spec,
```
Why are these made public, in contrast to their protected access mod settings in HSM?

istepic replied:
Done

johnfxgalea-e commented on src/mococrw/openssl_wrap.h:
```diff
@@ -1560,5 +1560,20 @@ SSL_EVP_PKEY_Ptr _ENGINE_load_public_key(ENGINE *e, const std::string &keyId);
+void _ENGINE_ctrl_cmd(ENGINE *e, const std::string &cmdName, void *p);
+
+/**
+ * Convert integer nid to curve name
```
You missed \p for nid parameter:

johnfxgalea-e replied:
@istepic I think you missed this one

istepic replied:
Not sure what you mean

istepic replied:
Done

johnfxgalea-e commented on tests/unit/test_hsm.cpp:
```diff
@@ -103,3 +110,43 @@ TEST_F(HSMTest, testHSMEngine)
+    auto engine = ::testutils::someEnginePtr();
+    auto pkey = ::testutils::somePkeyPtr();
+
+    EXPECT_CALL(_mock(), SSL_ENGINE_by_id(StrEq(engineID.c_str())))
```
Seems like we can reduce code duplication as a lot of the code is shared with HSMTest.

A helper function would be useful as I predict this code is going to be used also for other tests in the future.

istepic replied:
Done

johnfxgalea-e commented on tests/unit/test_key.cpp:
```diff
@@ -310,6 +310,51 @@ TEST_F(KeyHandlingTests, testKeyLoadPrivKeyFromHSM)
+                 MoCOCrWException);
+}
+
+TEST_F(KeyHandlingTests, testHSMKeyGenerationValidKeyId)
```
Nit: A test just like this but for RSA would be good for better coverage.

istepic replied:
Done

istepic committed:
fix John's comments

johnfxgalea-e commented on tests/unit/test_hsm.cpp:
```diff
@@ -79,7 +87,36 @@ void HSMTest::SetUp()
 }
 
-void HSMTest::TearDown() { openssl::OpenSSLLibMockManager::destroy(); }
+void HSMTest::TearDown()
```
Cannot you use this new helper `void HSMTest::initialiseEngine()` also for testHSMEngine now?

istepic replied:
Yeah. That makes testHSMEngine test redundant so I deleted it. initialiseEngine and testHSMEngine both test only the constructor.

johnfxgalea-e commented on src/mococrw/openssl_wrap.h:
```diff
@@ -1571,9 +1571,9 @@ void _ENGINE_ctrl_cmd(ENGINE *e, const std::string &cmdName, void *p);
 
 /**
- * Get number of bits key has from EVP_PKEY data structure
+ * Get number of bits key has from \p to EVP_PKEY data structure
```
I think you need to remove `to`, no?

istepic replied:
Done

johnfxgalea-e reviewed (changes requested):
Last few things

Commit:
add-fix

johnfxgalea-e reviewed (approved):
LGTM

garw commented on src/hsm.cpp:
```diff
@@ -35,14 +37,55 @@ HsmEngine::HsmEngine(const std::string &id, const std::string &modulePath, const
     return _ENGINE_load_private_key(_engine.get(), keyID);
 }
 
+openssl::SSL_EVP_PKEY_Ptr HsmEngine::generateKey(const RSASpec &spec,
```
this shouldn't be const even if it may work syntactically. Just from the idea: Creating a key changes the state of the HSM.

istepic replied:
Done

garw commented on src/mococrw/hsm.h:
```diff
@@ -40,19 +42,47 @@ class HSM
+     * @param keyLabel The key label
+     * @return AsymmetricKeypair @ref AsymmetricKeypair
+     */
+    virtual openssl::SSL_EVP_PKEY_Ptr generateKey(const RSASpec &spec,
```
This shouldn't be const even if it works syntactically here. Generating a key definately changes the state of the HSM.
Additionaly: Order of parameter. The key is placed into a token, so I would put the token-label first and then keyId/keylabel.

istepic replied:
Done

garw commented on src/mococrw/hsm.h:
```diff
@@ -40,19 +42,47 @@ class HSM
+     * @param keyLabel The key label
+     * @return AsymmetricKeypair @ref AsymmetricKeypair
+     */
+    virtual openssl::SSL_EVP_PKEY_Ptr generateKey(const ECCSpec &spec,
```
s.a.

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+    static AsymmetricKeypair readPrivateKeyFromHSM(const HSM &hsm, const std::string &keyID);
+
+    /**
+     * @brief Generates RSA keypair on HSM token according to the spec given. The keys are fetched
```
Only RSA? This looks now generic to me?

istepic replied:
Moved back to overloaded functions

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * @brief Generates RSA keypair on HSM token according to the spec given. The keys are fetched
+     * _only_ by key ID so generating multiple keys with the same ID must be avoided!
+     * @param hsm Initialized HSM engine handle
+     * @param spec @ref RSASpec
```
How about ECC spec?

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+                                              const std::string &tokenLabel,
+                                              const std::string &keyLabel)
+    {
+        static_assert(std::is_base_of<AsymmetricKey::Spec, T>::value,
```
Templates are bad for ABI stability and I wonder if it's really worth it here.

istepic replied:
Reverted to overloaded functions

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * _only_ by key ID so generating multiple keys with the same ID must be avoided!
+     * @param hsm Initialized HSM engine handle
+     * @param spec @ref RSASpec
+     * @param keyID ID of the key on HSM. We use IDs to fetch keys from HSM. _Only_ hex values are
```
Can we improve the documentation here a bit? This is one of the main entry points of a developer to the whole HSM setup, so it wouldn't hurt to explain some of the basics here: 
* IDs are the main identification mechanism we use (btw - why?, shouldn't we offer label as well?)
* IDs are internally big integers and we represent them here as hex strings
There is a vague statement about double IDs to be "avoided" what is this supposed to mean? What happens? Should we explicitly prevent this by checking if the ID is already taken?

istepic replied:
In theory we could offer a label identification as well. This is something we didn't really look into. Maybe because `key_id` is used for a function argument name that loads the keys but this `key_id` is obviously a more general concept.
EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                                  UI_METHOD *ui_method, void *callback_data);
                                  
We can talk about this but this should into seperate PR.

garw replied:
see also Ricardo's comment.

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * valid! Generating 2 keys with the same ID must be avoided. This is not prohibited by PKCS#11
+     * standard but since we currently only fetch by key ID, there is no way of ensuring that the
+     * correct key is fetched.
+     * @param tokenLabel Label of the token on HSM. This determines where they key shall be stored.
```
Same as above: Let's be a bit more generous regarding information here.
* One HSM may support multiple (separated) storage locations that we call token
* This parameter selects the token to put the key into

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+        if (keyID.length() > 127 || !std::all_of(keyID.begin(), keyID.end(), [](unsigned char c) {
+                return std::isxdigit(c);
+            })) {
+            throw MoCOCrWException(std::string("invalid keyID"));
```
More verbose error description: What does a valid keyID entail? Only a-f0-9, at max 127 letters. 
In addition we probably want to enforce that its an even amount of letters, otherwise this does not lead to full bytes (to mention also in documentation?)

istepic replied:
This is a bit weird yes. I think we shouldn't restrict users to use keyID's that have odd number of hex characters. Standard linux tool for managing HSMs, pkcs11-tool, adds 4 zeros on the 4 MSbs. I added a little bit of code that loads a key with the ID 0x0123 if "123" ID is given. And a little bit of code that creates a key with the ID 0x0123 if "123" is given.

I think this is the sanest approach as we rely less on the implementation of lower-level modules.

garw replied:
Now that I thought about it, I do agree with Ricardo tbh. std::string is not the right datatype here.

garw commented on src/mococrw/util.h:
```diff
@@ -23,7 +23,7 @@
 #include <openssl/crypto.h>
 
-#include "mococrw/error.h"
+#include "error.h"
```
Why?

istepic replied:
clangd was complaining for me

garw replied:
I don't think that this is correct though.

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * there
+     */
+    template <typename T>
+    static AsymmetricKeypair generateKeyOnHsm(const HSM &hsm,
```
Follow-up here: When the generation method won't be const anymore, it can't be const here.

istepic replied:
Done

garw commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,123 @@
+
+    /************** Get pubkey from HSM, store it in PEM file for later use **************/
+    /* Use-Case 2: The public key used for verification is stored in the HSM */
+    auto pubKeyEcc = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyIDEcc);
```
Do I really need to get it separately or could I just call publicKeyToPem() on the private key?

istepic replied:
This is done in use-case 3. I cleaned it up a bit.

garw commented on src/hsm.cpp:
```diff
@@ -17,6 +17,8 @@
  * #L%
  */
 #include "mococrw/hsm.h"
+#include "libp11.h"
```
Include order: Our headers first to prevent "hidden" compile errors because some definitions have been always implicitly been available to the header.
Among ours: preferably alphabetic if there is no technical reason against it?

istepic replied:
Done

istepic replied:
I changed this but clang-format prefers it this way for some reason.

garw commented on src/hsm.cpp:
```diff
@@ -35,14 +37,55 @@ HsmEngine::HsmEngine(const std::string &id, const std::string &modulePath, const
+                                                 const std::string &tokenLabel,
+                                                 const std::string &keyLabel) const
+{
+    PKCS11_RSA_KGEN pkcs11_rsa_spec;
```
naming convention of our code is camelCase.

istepic replied:
Done

garw commented on src/hsm.cpp:
```diff
@@ -35,14 +37,55 @@ HsmEngine::HsmEngine(const std::string &id, const std::string &modulePath, const
+    PKCS11_EC_KGEN pkcs11_ec_spec;
+    std::string curve{};
+    try {
+        curve = _EC_curve_nid2nist(int(spec.curve()));
```
This could now be just `spec.curveName()`?!

istepic replied:
Done

garw commented on src/hsm.cpp:
```diff
@@ -35,14 +37,55 @@ HsmEngine::HsmEngine(const std::string &id, const std::string &modulePath, const
+                                                 const std::string &keyLabel) const
+{
+    PKCS11_EC_KGEN pkcs11_ec_spec;
+    std::string curve{};
```
```suggestion
    std::string curve;
```
We just don't have this style in our code usually.

istepic replied:
Done

garw commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,123 @@
+    /* Read public key from HSM */
+    auto pubKeyRsa = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyIDRsa);
+
+    /* Do whatever you want using the private and public key. See rsa-example.cpp or
```
Considering that running the examples is the simplest form of integration test that we so/have. Add some actual example code here even if it is not really required to demonstrate anything?

istepic replied:
Done

garw commented on src/mococrw/hsm.h:
```diff
@@ -40,19 +42,47 @@ class HSM
+    /**
+     * @brief Generate a RSA key pair on the HSM
+     *
+     * @param spec The RSA specification @ref RSASpec
```
Not sure how much value is in super repetitive param descriptions. If they speak for themselves, just drop them. Otherwise, see that some value is actually added.

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+    /**
+     * @brief Generates RSA keypair on HSM token according to the spec given. The keys are fetched
+     * _only_ by key ID so generating multiple keys with the same ID must be avoided!
+     * @param hsm Initialized HSM engine handle
```
What is initialized supposed to mean in this context here? Does this refer to some state of the HSM object? What does the caller need to do? If nothing, just drop it to avoid confusion.

istepic replied:
Done

garw commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,51 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+            return AsymmetricKeypair{std::move(key)};
+        } catch (const openssl::OpenSSLException &e) {
+            throw MoCOCrWException(
+                    std::string("Key generation failed for unknown reason. Likely reasons: invalid "
```
Not sure if the "likely reasons" are actually helpful. Feels a bit like just blaming others while we don't actually know anything helpful.

istepic replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1597,5 +1597,22 @@ void _ENGINE_finish(ENGINE *e)
+std::string _EC_curve_nid2nist(int nid)
+{
+    auto nist = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EC_curve_nid2nist, nid);
+    return std::string(nist);
```
minor: `std::string()` shouldn't be required here. The constructor from `const char*` isn't explicit.

istepic replied:
Done

garw commented on tests/unit/test_hsm.cpp:
```diff
@@ -74,32 +83,45 @@ void HSMTest::SetUp()
+    auto hsm = initialiseEngine();
+    EXPECT_CALL(_mock(), SSL_EC_curve_nid2nist(curve)).WillOnce(Return("P-256"));
+    EXPECT_CALL(_mock(),
+                SSL_ENGINE_ctrl_cmd(engine, StrEq("KEYGEN"), 0 /*non-optional*/, _, nullptr, 1))
```
If we would want to be more thorough, we would also match the struct that is passed to the engine command. But I guess for now its ok to have it like this. In particular, given that engines are deprecated and to be replaced by providers anyways.

garw commented on tests/unit/test_key.cpp:
```diff
@@ -373,6 +421,18 @@ TEST_F(KeyHandlingTests, testGetKeySpec)
+    EXPECT_EQ(dynamic_cast<ECCSpec *>(defaultSpec.get())->curveName(), "P-256");
+    EXPECT_EQ(dynamic_cast<ECCSpec *>(Sect571r1Spec.get())->curveName(), "B-571");
+    EXPECT_EQ(dynamic_cast<ECCSpec *>(Secp521Spec.get())->curveName(), "P-521");
+    /* These two curves are not working:
```
This is to be expected as the Ed curves have a slightly different implementation in OpenSSL. Our implementations also look different. We shouldn't worry to much, or think about a better way to implement this function then…

garw commented on tests/unit/test_opensslwrapper.cpp:
```diff
@@ -340,3 +340,41 @@ TEST_F(OpenSSLWrapperTest, testEngineFinishThrowsException)
     EXPECT_THROW(_ENGINE_finish(engine), OpenSSLException);
 }
+
+TEST_F(OpenSSLWrapperTest, testEngineCtrlCmd)
```
I don't really think that we need tests for the wrap layer. At least we usually don't have this if the function just implements error handling using the CallChecked templates.

istepic replied:
Agreed but I guess these don't hurt much as well

garw reviewed (changes requested):
Overall looks good. Yet, there is a list of minor issues to iron out.

istepic committed:
address garw's comments

istepic committed:
address garw's comments

dfuhrmann commented on src/mococrw/key.h:
```diff
@@ -270,7 +271,55 @@ class AsymmetricKeypair : public AsymmetricPublicKey
      * object as a result.
      */
-    static AsymmetricKeypair readPrivateKeyFromHSM(HSM &hsm, const std::string &keyID);
+    static AsymmetricKeypair readPrivateKeyFromHSM(const HSM &hsm, const std::string &keyID);
```
Why is the key identified by tokenLabel, keyID and keyLabel in the generation API, but here only KeyID is used?

Would be cool to consolidate and minimize the number of artifacts you need to identify one key pair.

istepic replied:
Good point. Key ID is the only artifact you need to identify a key pair. I guess adding keyLabel in generation API generates some confusion and we will probably create another PR with the intention to eliminate that confusion.

ricleite commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,202 @@
     // Information for engine loading and key management.
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
     std::string pin("1234");
```
heh, should applications really hardcode this instead of `utility::cryptoRandomBytes`?

(@garw that would... ease some burdens ;) )

istepic replied:
Added a comment to  warn people not to hardcore pins

ricleite commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,55 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * are stored. This determines on which token the key shall be stored.
+     * @param keyLabel Arbitrary key label
+     * @return AsymmetricKeypair @ref AsymmetricKeypair
+     * @throw MoCOCrWException Since most of the logic is happening outside of OpenSSL and inside
+     * libp11 and HSM module implementation, we can't know exactly what went wrong. libp11 does log
+     * some things to stderr, check if there's more context there
```
uff! not good

I would ask that the exceptions' `.what()` states the same thing that's here, as other mococrw exceptions always have useful info

(it's also a bit weird that libmococrw now does stdout/stderr due to this...)

istepic replied:
Yeah, this is unfortunate but there's not much we can do about it as error handling between 3 independently developed libraries is always a mess. Not much is logged to stderr by libp11 though.

ricleite commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,55 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * current libp11 implementation.
+     * @param hsm HSM engine handle
+     * @param spec @ref RSASpec
+     * @param keyID ID of the key on HSM. _Only_ hex values are valid because key IDs are stored as
+     * raw bytes.
```
if only hex values are valid anyway, perhaps this shouldn't be an `std::string`

is there any hard reason to keep it like this? it is very easy as a user to make a mistake

garw replied:
I kind of agree that this then maybe should better be a std::vector<uint8_t>

dfuhrmann replied:
@garw I understand that the HSM expect a certain amount of raw bytes as keyID. But for the user of this interface, I think it would be most convenient to be able to also pass a human readable string as a keyID, as this can be more easily handled / used on caller site.

Why not keep std::string as a data type then, and just change semantics? This data type is perfectly fine for both passing raw bytes, as well as (ASCII / UTF8) strings. So the caller could decide to pass a human readable string, and internally you just check for a maximum size of the data, and treat any bytes as raw data.

istepic replied:
This important point will be handled in a separate PR. I don't want it to get lost in this already huge PR so stay tuned.

dfuhrmann replied:
All fine, thanks for the consideration.

ricleite commented on src/mococrw/key.h:
```diff
@@ -270,7 +270,55 @@ class AsymmetricKeypair : public AsymmetricPublicKey
+     * current libp11 implementation.
+     * @param hsm HSM engine handle
+     * @param spec @ref RSASpec
+     * @param keyID ID of the key on HSM. _Only_ hex values are valid because key IDs are stored as
+     * raw bytes.
```
no limitations on the key size or encoding otherwise? weird, I'd expect an HSM to be a lot more picky

garw replied:
This is probably implementation specific but not mentioned in the general PKCS#11 spec

ricleite commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,202 @@
 {
     // Information for engine loading and key management.
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
```
TIL OpenSSL can `dlopen` stuff... interesting

ricleite commented on examples/hsm-example.cpp:
```diff
@@ -26,20 +28,202 @@
-
-    // Step 2: Initialise the PKCS11 Engine.
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
     HsmEngine hsmEngine(id, modulePath, pin);
```
hm, the pin seems to be associated to the HSM, instead of the key itself

hypothetical scenario... how could an `application A` create a key in the HSM, and share it only with an `application B`, such that no `application C` (even if it can access the HSM as well) can use it?

it does not seem possible with this mococrw design, unless I'm missing something

garw replied:
I discussed the same thing with Ivan today. We'll move the "token" also into the HSMEngine. Then one HSMEngine represents on PKCS#11 token instead of the whole HSM. Then everything that belongs together token/PIN is in the same thing. This is not really clear from the current design.

istepic replied:
This will be handled in a separate PR. The scenario you described currently doesn't really work as you would expect.

ricleite reviewed (commented):
How/where to do HSM initialization? Or is this out of scope for mococrw?

garw commented on examples/hsm-example.cpp:
```diff
@@ -100,51 +163,67 @@ int main(void)
-    auto rsaPrivKey = AsymmetricPrivateKey::generateKeyOnHsm(
-            hsmEngine, rsaSpec, keyIDRsa, "token-label", "BarfoKey");
+    auto rsaSignatureDigestType = DigestTypes::SHA512;
+    auto mgf1 = std::make_shared<MGF1>(MGF1(mgf1DigestType));
```
make_shared is a fowarding constructor. You shouldn't create the object itself in it. Otherwise, you don't really get the benefits of it and the object will most probably copy constructed/move constructed from your temporary object.

istepic replied:
Done

garw commented on examples/hsm-example.cpp:
```diff
@@ -100,51 +163,67 @@ int main(void)
+    auto rsaSignatureDigestType = DigestTypes::SHA512;
+    auto mgf1 = std::make_shared<MGF1>(MGF1(mgf1DigestType));
+    int saltLength = mococrw::Hash::getDigestSize(rsaSignatureDigestType);
+    auto padding = std::make_shared<PSSPadding>(PSSPadding(mgf1, saltLength));
```
s.a.

istepic replied:
done

garw commented on examples/hsm-example.cpp:
```diff
@@ -100,51 +163,67 @@ int main(void)
+     */
+    auto RSAsignature = rsaSign(rsaPrivKey, rsaSignatureDigestType, padding, message);
+    /* We can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
```
Isn't is usually called generalitzation if you do an upcast? Anyways, I don't really think that this comment is needed. Downcasts are never implicit and need to be explicitly stated.

istepic replied:
Done

garw reviewed (changes requested):
Let's briefly look through the remaining few findings and David's/Ricardo's comments tomorrow.

istepic committed:
Merge pull request #4 from istepic/hsm-keygen

Add hsm keygen

istepic committed:
hsm: associate each HsmEngine object with a token

This prevents using HsmEngine object with set pin
from trying to load/generate keys from/on other tokens

garw commented on src/mococrw/util.h:
```diff
@@ -45,6 +45,15 @@ std::vector<uint8_t> fromHex(const std::string &hexData);
+ * are used in percent encoding). If you have a different use case, change this or create your
+ * own function.
+ */
+void hexStringToPctEncoded(std::string &str);
```
I don't think that this is a generally useful function. In URIs only non-ascii characters are escaped via this ampersand notation. The function here does no particular encoding but just adds an % every second character.  We should have this as a private function. Moreover, functions with side-effects are generally discouraged. This should be a string -> string function.

garw reviewed (changes requested):
mostly ok now w.r.t. what we discussed. Just this one utility function is a bit odd and shouldn't become part of our public interface.

istepic committed:
Associate HsmEngine with token

garw reviewed (approved):
lg2m now
